crates/bevy_time/src/time.rs

```diff
@@ -1,218 +1,288 @@
-
-        self.elapsed_wrapped = duration_div_rem(self.elapsed, self.wrap_period).1;
+        self.elapsed_wrapped = duration_rem(self.elapsed, self.wrap_period);
         self.elapsed_seconds_wrapped = self.elapsed_wrapped.as_secs_f32();
         self.elapsed_seconds_wrapped_f64 = self.elapsed_wrapped.as_secs_f64();
-        self.raw_elapsed_wrapped = duration_div_rem(self.raw_elapsed, self.wrap_period).1;
-        self.raw_elapsed_seconds_wrapped = self.raw_elapsed_wrapped.as_secs_f32();
-        self.raw_elapsed_seconds_wrapped_f64 = self.raw_elapsed_wrapped.as_secs_f64();
-
-        self.last_update = Some(instant);
     }
 
-    /// Returns the [`Instant`] the clock was created.
+    /// Advance this clock to a specific `elapsed` time.
     ///
-    /// This usually represents when the app was started.
-    #[inline]
-    pub fn startup(&self) -> Instant {
-        self.startup
+    /// [`Self::delta()`] will return the amount of time the clock was advanced
+    /// and [`Self::elapsed()`] will be the `elapsed` value passed in. Cannot be
+    /// used to move time backwards.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `elapsed` is less than `Self::elapsed()`.
+    pub fn advance_to(&mut self, elapsed: Duration) {
+        assert!(
+            elapsed >= self.elapsed,
+            "tried to move time backwards to an earlier elapsed moment"
+        );
+        self.advance_by(elapsed - self.elapsed);
     }
 
-    /// Returns the [`Instant`] when [`update`](#method.update) was first called, if it exists.
+    /// Returns the modulus used to calculate [`elapsed_wrapped`](#method.elapsed_wrapped).
     ///
-    /// This usually represents when the first app update started.
+    /// **Note:** The default modulus is one hour.
     #[inline]
-    pub fn first_update(&self) -> Option<Instant> {
-        self.first_update
+    pub fn wrap_period(&self) -> Duration {
+        self.wrap_period
     }
 
-    /// Returns the [`Instant`] when [`update`](#method.update) was last called, if it exists.
+    /// Sets the modulus used to calculate [`elapsed_wrapped`](#method.elapsed_wrapped).
+    ///
+    /// **Note:** This will not take effect until the next update.
+    ///
+    /// # Panics
     ///
-    /// This usually represents when the current app update started.
+    /// Panics if `wrap_period` is a zero-length duration.
     #[inline]
-    pub fn last_update(&self) -> Option<Instant> {
-        self.last_update
+    pub fn set_wrap_period(&mut self, wrap_period: Duration) {
+        assert!(!wrap_period.is_zero(), "division by zero");
+        self.wrap_period = wrap_period;
     }
 
-    /// Returns how much time has advanced since the last [`update`](#method.update), as a [`Duration`].
+    /// Returns how much time has advanced since the last [`update`](#method.update), as a
+    /// [`Duration`].
     #[inline]
     pub fn delta(&self) -> Duration {
         self.delta
     }
```

Comment:
This does not seem usable well when updates are done asynchronously. I want a function like [`Instant::duration_since`](https://doc.rust-lang.org/std/time/struct.Instant.html#method.duration_since).

Reply:
I talked in Discord how time could be tracked per system – so `time.delta()` in that system would calculate the delta since the last time that system was run. However, that was deemed too complex. 

Explicitly tracking time deltas requires the user to store the previous `time.elapsed()` value somewhere. `Instant::duration_since` equivalent is simply `time.elapsed() - since`. `Time::elapsed` is the `Instant` equivalent inside Bevy, one that isn't platform specific mess like Instant, and `duration_since` is just substraction. Also `checked_sub` and `saturating_sub` are available for Durations if there's a need for specific underflow behaviour.

Reply:
> I talked in Discord how time could be tracked per system – so `time.delta()` in that system would calculate the delta since the last time that system was run. However, that was deemed too complex.

For a few reasons.

1. We can't actually auto-update `Time` with this per-system value. `Time` can't even be used inside `bevy_ecs` because `bevy_time` depends on `bevy_ecs`.
2. Using a unique `Δt` per system is not even a valid way to intergrate time. The same sequence of logic would get different results depending on how many systems it was split into. It's super suspect that different components on the same entity could be integrated by different amounts of time.
3. It's too different from how everyone knows "delta time". It'd break unity with fixed update, which is counterproductive.

Reply:
Maybe that's right. This was probably not good idea.

Reply:
Per system delta time that will track delta time from the last time that specific system was run:
```rust
fn my_system(time: Res<Time>, mut last: Local<Duration>) {
    let delta = time.elapsed() - last;
    last = time.elapsed();
    ...
}
```
Could probably make a SystemParam and implement some convenience method on it to make it even nicer to use.

But I do believe we want to keep `time.delta()` as is, as that matches what people expect from it.